python/adaptivity.py.in

```diff
@@ -1028,47 +1085,52 @@ def mesh_metric(mesh):
   #NOTE: For stabilising advective problems where the local velocity direction 
   #      is vdir (unit length), the local element size, Lh, can be taken as 
   #      htmp = dot(vdir,M); Lh = 1/sqrt(dot(htmp,htmp)), where M is the mesh metric
+  dim = mesh.geometry().dim()
   cell2dof = c_cell_dofs(mesh,TensorFunctionSpace(mesh, "DG", 0))
-  cells = mesh.cells()
+  cells_array = mesh.cells()
   coords = mesh.coordinates()
-  p1 = coords[cells[:,0],:];
-  p2 = coords[cells[:,1],:];
-  p3 = coords[cells[:,2],:];
+  n_cells = mesh.num_cells()
+  p1 = coords[cells_array[:,0],:];
+  p2 = coords[cells_array[:,1],:];
+  p3 = coords[cells_array[:,2],:];
   r1 = p1-p2; r2 = p1-p3; r3 = p2-p3
   Nedg = 3
-  if mesh.geometry().dim() == 3:
+  if dim == 3:
       Nedg = 6
-      p4 = coords[cells[:,3],:];
+      p4 = coords[cells_array[:,3],:];
       r4 = p1-p4; r5 = p2-p4; r6 = p3-p4
-  rall = zeros([p1.shape[0],p1.shape[1],Nedg])
+  rall = zeros([n_cells,dim,Nedg])
   rall[:,:,0] = r1; rall[:,:,1] = r2; rall[:,:,2] = r3;
-  if mesh.geometry().dim() == 3:
+  if dim == 3:
       rall[:,:,3] = r4; rall[:,:,4] = r5; rall[:,:,5] = r6
-  All = zeros([p1.shape[0],Nedg**2])
+  All = zeros([n_cells,Nedg**2])
   inds = arange(Nedg**2).reshape([Nedg,Nedg])
   for i in range(Nedg):
     All[:,inds[i,0]] = rall[:,0,i]**2; All[:,inds[i,1]] = 2.*rall[:,0,i]*rall[:,1,i]; All[:,inds[i,2]] = rall[:,1,i]**2
-    if mesh.geometry().dim() == 3:
+    if dim == 3:
       All[:,inds[i,3]] = 2.*rall[:,0,i]*rall[:,2,i]; All[:,inds[i,4]] = 2.*rall[:,1,i]*rall[:,2,i]; All[:,inds[i,5]] = rall[:,2,i]**2
-  Ain = zeros([Nedg*2-1,Nedg*p1.shape[0]])
-  ndia = zeros(Nedg*2-1)
-  for i in range(Nedg):
-      for j in range(i,Nedg):
-          iks1 = arange(j,Ain.shape[1],Nedg)
-          if i==0:
-              Ain[i,iks1] = All[:,inds[j,j]]
-          else:
-              iks2 = arange(j-i,Ain.shape[1],Nedg)
-              Ain[2*i-1,iks1] = All[:,inds[j-i,j]]
-              Ain[2*i,iks2]   = All[:,inds[j,j-i]]
-              ndia[2*i-1] = i
-              ndia[2*i]   = -i
-    
-  A = scipy.sparse.spdiags(Ain, ndia, Ain.shape[1], Ain.shape[1]).tocsr()
-  b = ones(Ain.shape[1])
-  X = scipy.sparse.linalg.spsolve(A,b)
+  #Ain = zeros([Nedg*2-1,Nedg*n_cells])
+  #ndia = zeros(Nedg*2-1)
+  #for i in range(Nedg):
+  #    for j in range(i,Nedg):
+  #        iks1 = arange(j,Ain.shape[1],Nedg)
+  #        if i==0:
+  #            Ain[i,iks1] = All[:,inds[j,j]]
+  #        else:
+  #            iks2 = arange(j-i,Ain.shape[1],Nedg)
+  #            Ain[2*i-1,iks1] = All[:,inds[j-i,j]]
+  #            Ain[2*i,iks2]   = All[:,inds[j,j-i]]
+  #            ndia[2*i-1] = i
+  #            ndia[2*i]   = -i
+  #  
+  #A = scipy.sparse.spdiags(Ain, ndia, Ain.shape[1], Ain.shape[1]).tocsr()
+  #b = ones(Ain.shape[1])
+  #X = scipy.sparse.linalg.spsolve(A,b)
+
+  X = numpy.hstack([numpy.linalg.solve(item.reshape((Nedg,Nedg)),ones(Nedg)) for item in All])
```

Comment:
Ok to be honest, I don't understand what the previous version was trying to do, and what this is doing. Can you please explain/comment ?